File: src/week3.reduce-array-to-the-half/papakang22.java
import java.util.Map;
import java.util.Map.Entry;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

// Runtime: 42 ms, faster than 40.74% of Java online submissions for Reduce Array Size to The Half.
// Memory Usage: 55.2 MB, less than 28.70% of Java online submissions for Reduce Array Size to The Half.
class papakang22 {
    public int minSetSize(int[] arr) {
        Integer half = arr.length % 2 == 0 ? arr.length / 2 : arr.length / 2 + 1;
        
        Map<Integer,Integer> map = new HashMap<Integer,Integer>();
        for(int num : arr){ //맵에다 개수 취합
            Integer cnt;
            if(map.get(num) != null){
                cnt = map.get(num) + 1;
            }else{
                cnt = 1;
            }  
            map.put(num,cnt);
        }
        
        List<Entry<Integer, Integer>> list_entries = new ArrayList<Entry<Integer, Integer>>(map.entrySet());

		// 비교함수 Comparator를 사용하여 내림차순으로 정렬
		Collections.sort(list_entries, new Comparator<Entry<Integer, Integer>>() {
			// compare로 값을 비교
			public int compare(Entry<Integer, Integer> obj1, Entry<Integer, Integer> obj2) {
				// 내림 차순 정렬
				return obj2.getValue().compareTo(obj1.getValue());
			}
		});
        
        Integer hap = 0; //없어질 개수
        List<Integer> answer = new ArrayList<Integer>();
        for(Entry<Integer, Integer> entry : list_entries) {
			Integer key = entry.getKey();
            Integer val = entry.getValue();
            hap += val;
            answer.add(key);
            
            if(half <= hap){ //없어질 갯수가 배열길이의 반 이상일 경우 break
                break;
            }
		}
    
        return answer.size();
    }
}

File: src/week10.h-index/papakang22.js
// 테스트 1 〉	통과 (0.36ms, 30.1MB)
// 테스트 2 〉	통과 (0.29ms, 29.9MB)
// 테스트 3 〉	통과 (0.47ms, 30.1MB)
// 테스트 4 〉	통과 (0.43ms, 30.1MB)
// 테스트 5 〉	통과 (0.52ms, 29.7MB)
// 테스트 6 〉	통과 (0.57ms, 30.1MB)
// 테스트 7 〉	통과 (0.28ms, 30.3MB)
// 테스트 8 〉	통과 (0.11ms, 30.1MB)
// 테스트 9 〉	통과 (0.13ms, 30MB)
// 테스트 10 〉	통과 (0.31ms, 30.1MB)
// 테스트 11 〉	통과 (0.64ms, 30.1MB)
// 테스트 12 〉	통과 (0.14ms, 30MB)
// 테스트 13 〉	통과 (0.59ms, 30MB)
// 테스트 14 〉	통과 (0.54ms, 30.1MB)
// 테스트 15 〉	통과 (0.56ms, 30MB)
// 테스트 16 〉	통과 (0.09ms, 30.1MB)
function solution(citations) {
    var answer = -1;
    citations.sort((a,b) =>{
        return b-a;
    });
    
    for(let i = 0 ; i < citations.length; i++){
        if(citations[i] < i + 1){
            answer = i;
            break;
        }
    }
    
    return answer > -1 ? answer : citations.length;
}